lib/features/docker_save.js

```diff
@@ -9,41 +11,100 @@ import zlib from 'zlib';
 
 let debug = Debug('docker-worker:features:docker-save');
 
-const ARTIFACT_NAME = 'public/dockerImage.tar'
+//must be changed if the delimiter in lib/volume_cache.js is changed
+let KEY_DELIMITER = '::';
 
 export default class DockerSave {
   async killed (task) {
-    //temporary path for saved file
-    let pathname = path.join('/tmp', slugid.v4() + '.tar.gz');
-
-    let {Id: imageId} = await task.dockerProcess.container.commit({
-      repo: 'task/' + task.status.taskId + '/' + task.runId
-    });
-    let image = task.runtime.docker.getImage('task/' + task.status.taskId + '/' + task.runId + ':latest');
-    let imgStream = await image.get();
-    let zipStream = zlib.createGzip();
-    imgStream.pipe(zipStream).pipe(fs.createWriteStream(pathname));
-    await waitForEvent(zipStream, 'end');
-    debug('tar written');
-
-    let stat = await fs.stat(pathname);
-    debug(stat.size);
-    let uploadStream = fs.createReadStream(pathname);
-    let expiration = new Date(Date.now() + task.runtime.dockerSave.expiration);
-
-    await uploadToS3(task, uploadStream, ARTIFACT_NAME, expiration, {
-      'content-type': 'application/x-tar',
-      'content-length': stat.size,
-      'content-encoding': 'gzip'
-    });
-
-    debug('artifact uploaded');
-
-    //cleanup
-    fs.unlink(pathname).catch(() => {
-      task.runtime.log('[alert-operator] could not delete docker save tarball, worker may run out of hdd space');
-    });
-    await image.remove();
+    let errors = [];
+    let volumeCaches = task.volumeCaches || [];
+    await Promise.all(volumeCaches.map(async (cacheStr) => {
+      //saving the caches
+      try {
+        let cache = cacheStr.split(KEY_DELIMITER);
+        let loc = path.join(task.runtime.cache.volumeCachePath, cache[0], cache[1] + '/');
+
+        //temporary path for saved file
+        let pathname = path.join('/tmp', slugid.v4() + '.tar.gz');
+        let zipStream = tar.pack(loc, { dereference: true }).pipe(zlib.createGzip());
+        zipStream.pipe(fs.createWriteStream(pathname));
+        await new Promise((accept, reject) => {
+          zipStream.on('end', accept);
+          zipStream.on('error', (error) => reject(error));
+        });
+        let expiration = taskcluster.fromNow(task.runtime.dockerSave.expiration);
+        let stat = await fs.stat(pathname);
+
+        await uploadToS3(task,
+          fs.createReadStream(pathname),
+          'public/cache/' + cache[0] + '.tar.gz',
```

Comment:
My bad... when I think I told you to use `.tar.gz`, I saw you changes in the test and realize `.tar` is sufficient because when using `content-encoding: gzip` then the HTTP layer (curl, wget, superagent, etc) will do the gzip decompression for you. It's suddenly acts as a transport layer encoding instead of being the encoding of our downloaded result..

@gregarndt, ^? crazy or reasonably sane... I think we do this with logs too... We don't append `.gz` because the HTTP client handles gzip when S3 serves with `content-encoding: gzip`